crates/bevy_ecs/src/schedule/schedule.rs

```diff
@@ -300,6 +309,92 @@ impl Schedule {
             system.apply_buffers(world);
         }
     }
+
+    /// Get the state of system-stepping
+    pub fn stepping(&self) -> bool {
+        match self.executable.step_state {
+            StepState::RunAll => false,
+            StepState::Wait(_) | StepState::Next(_) | StepState::Remaining(_) => true,
+        }
+    }
+
+    /// When stepping mode is enabled, this method will return the name of the
+    /// next system to be run when the schedule is stepped.
+    pub fn next_step_system_name(&self) -> Option<std::borrow::Cow<'static, str>> {
+        let index = self.executable.stepping_next_system_index()?;
+        Some(self.executable.systems[index].name())
+    }
+
+    /// When stepping mode is enabled, this method will return the [`NodeId`]
+    /// of the next system to be run when this schedule is stepped.
+    pub fn next_step_system_node_id(&self) -> Option<NodeId> {
+        let index = self.executable.stepping_next_system_index()?;
+        self.executable.system_ids.get(index).copied()
+    }
+
+    /// Check if a given system supports stepping
+    ///
+    /// This function will return false for an unknown `id`, or when the
+    /// schedule has not yet been built by calling [`Schedule::run()`].
+    pub fn system_permits_stepping(&self, id: NodeId) -> bool {
+        self.executable
+            .system_ids
+            .iter()
+            .position(|&i| i == id)
+            .map(|index| self.executable.systems_with_stepping_enabled[index])
+            .unwrap_or(false)
+    }
+
+    /// Get a System by [`NodeId`]
+    pub fn system_at(&self, id: NodeId) -> Option<&BoxedSystem> {
```

Comment:
For these methods that access the `SystemSchedule`, if these are called before the `Schedule` is run, the returned value will be `None` because the `SystemSchedule` isn't built until `Schedule::run()` is called.

Now, it's not possible for us to build the `SystemSchedule` when these are called because `ScheduleGraph::build_schedule` requires `Components` from `World`.  I'd prefer not to add that requirement to all of these getter functions.

@alice-i-cecile Is the appropriate path to switch the return type to `Result<Option<...>>`, and return some equivalent of `EAGAIN` to allow callers to differentiate between "not ready", and "not there"?

Reply:
Done, `Result<Option<...>, ScheduleNotBuiltError>`